helix-term/src/commands.rs

```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
 }
 
 // I inserts at the first nonwhitespace character of each line with a selection
+// If the line is empty, automatically indent
 fn insert_at_line_start(cx: &mut Context) {
-    goto_first_nonwhitespace(cx);
     enter_insert_mode(cx);
+
+    let (view, doc) = current!(cx.editor);
+
+    let text = doc.text().slice(..);
+    let contents = doc.text();
+    let selection = doc.selection(view.id);
+
+    let language_config = doc.language_config();
+    let syntax = doc.syntax();
+    let tab_width = doc.tab_width();
+
+    let mut ranges = SmallVec::with_capacity(selection.len());
+    let mut offs = 0;
+
+    let mut transaction = Transaction::change_by_selection(contents, selection, |range| {
+        let cursor = range.from();
+        let cursor_line = text.char_to_line(cursor);
+
+        let doc_end = contents.len_chars();
+        if range.from() == doc_end {
+            ranges.push(Range::point(doc_end + offs));
+            return (doc_end, doc_end, None);
+        }
+
+        let line_content = line_without_line_ending(&text, cursor_line);
+        if line_content.len_chars() == 0 {
+            // line is empty => auto indent
+            let (line_end_index, line_end_offset_width) = if cursor_line == 0 {
+                (0, 0)
+            } else {
+                (
+                    line_end_char_index(&doc.text().slice(..), cursor_line),
```

Comment:
```suggestion
                    cursor_line_start
```
Avoid recalculating that here the line is always empty (you checked above), so the result of `line_end_char_index` is always the same as the line start

Reply:
Done.